@claude fix:
- lint: secator/cli.py:1273:121: E501 line too long (123 > 120 characters)
- unit: 
```
FAILED tests/unit/test_query_utils.py::TestValidateQueryFields::test_invalid_field_removed_with_warning - AssertionError: Expected 'print' to have been called once. Called 0 times.
FAILED tests/unit/test_query_utils.py::TestValidateQueryFields::test_or_validates_each_fragment - AssertionError: assert {'$or': [{'_type': 'url', 'status_code': {'$ne': 200}}, {}]} == {'_type': 'url', 'status_code': {'$ne': 200}}
  
  Left contains 1 more item:
  {'$or': [{'_type': 'url', 'status_code': {'$ne': 200}}, {}]}
  Right contains 2 more items:
  {'_type': 'url', 'status_code': {'$ne': 200}}
  
  Full diff:
    {
  +     '$or': [
  +         {
  -     '_type': 'url',
  +             '_type': 'url',
  ? ++++++++
  -     'status_code': {
  +             'status_code': {
  ? ++++++++
  -         '$ne': 200,
  +                 '$ne': 200,
  ? ++++++++
  +             },
  +         },
  +         {},
  -     },
  ?     ^
  +     ],
  ?     ^
    }
FAILED tests/unit/test_query_utils.py::TestValidateQueryFields::test_or_with_partial_invalid_fields - AssertionError: Expected 'print' to have been called once. Called 0 times.
FAILED tests/unit/test_query_utils.py::TestValidateQueryFields::test_and_validates_each_fragment - AssertionError: assert {'$and': [{'_context.scan_id': '5'}, {}]} == {'_context.scan_id': '5'}
  
  Left contains 1 more item:
  {'$and': [{'_context.scan_id': '5'}, {}]}
  Right contains 1 more item:
  {'_context.scan_id': '5'}
  
  Full diff:
    {
  +     '$and': [
  +         {
  -     '_context.scan_id': '5',
  +             '_context.scan_id': '5',
  ? ++++++++
  +         },
  +         {},
  +     ],
    }
FAILED tests/unit/test_query_utils.py::TestValidateQueryFields::test_multiple_invalid_fields_all_warned - AssertionError: assert 0 == 2
 +  where 0 = <MagicMock name='console.print' id='140647320688720'>.call_count
 +    where <MagicMock name='console.print' id='140647320688720'> = <MagicMock name='console' id='140647320689008'>.print
```

--- a/secator/cli.py
+++ b/secator/cli.py
@@ -1270,7 +1270,9 @@ def run_report_show(report_query, output, time_delta, query, fmt, workspace, dri
 
 	REPORT_QUERY: comma-separated runner paths (e.g. scans/5,tasks/3).
 	"""
-	from secator.query.utils import parse_report_paths, python_expr_to_mongo, validate_query_fields, query_has_type_constraint
+	from secator.query.utils import (
+		parse_report_paths, python_expr_to_mongo, validate_query_fields, query_has_type_constraint
+	)
 
 	current = get_file_timestamp()
 	workspace_name = workspace or CONFIG.workspace.default or 'default'

--- a/secator/query/utils.py
+++ b/secator/query/utils.py
@@ -361,6 +361,87 @@ def python_expr_to_mongo(query):
     return result
 
 
+def _warn_unknown_field(field_name, type_name, valid_fields):
+    """Emit a warning for an unknown field in a query fragment."""
+    from secator.output_types import Warning
+    public_fields = sorted(f for f in valid_fields if not f.startswith('_'))
+    console.print(Warning(
+        message=f"Field '{field_name}' does not exist on type '{type_name}'. "
+                f"Available fields: {', '.join(public_fields)}"
+    ))
+
+
+def _build_type_map():
+    from secator.output_types import OUTPUT_TYPES
+    return {
+        cls.__dataclass_fields__['_type'].default: cls
+        for cls in OUTPUT_TYPES
+        if '_type' in getattr(cls, '__dataclass_fields__', {})
+        and isinstance(cls.__dataclass_fields__['_type'].default, str)
+    }
+
+
+def _validate_fragment(fragment, type_map):
+    """Validate a single query fragment against known output type fields.
+
+    Returns None if all user-specified fields were invalid (signals the caller
+    to drop this fragment from an $or/$and list rather than keeping a bare
+    {'_type': 'x'} that would match every object of that type).
+    """
+    if not isinstance(fragment, dict):
+        return fragment
+    _type = fragment.get('_type')
+    if not _type:
+        return fragment
+    cls = type_map.get(_type)
+    if cls is None:
+        return fragment
+    valid_fields = set(cls.fields())
+    result = {}
+    user_fields_seen = 0
+    user_fields_kept = 0
+    for k, v in fragment.items():
+        if k.startswith('$') or k.startswith('_'):
+            result[k] = v
+            continue
+        user_fields_seen += 1
+        top_field = k.split('.')[0]
+        if top_field in valid_fields:
+            result[k] = v
+            user_fields_kept += 1
+        else:
+            _warn_unknown_field(k, _type, valid_fields)
+    # If the fragment had user-specified fields but ALL were invalid, signal
+    # the caller to drop this fragment entirely rather than keeping a bare
+    # {'_type': _type} that would match every object of that type.
+    if user_fields_seen > 0 and user_fields_kept == 0:
+        return None
+    return result
+
+
+def _validate_query(q, type_map):
+    if not isinstance(q, dict):
+        return q
+    if '$or' in q:
+        parts = [_validate_query(sub, type_map) for sub in q['$or']]
+        parts = [p for p in parts if p]
+        if len(parts) == 0:
+            return {}
+        if len(parts) == 1:
+            return parts[0]
+        return {'$or': parts}
+    if '$and' in q:
+        parts = [_validate_query(sub, type_map) for sub in q['$and']]
+        parts = [p for p in parts if p]
+        if len(parts) == 0:
+            return {}
+        if len(parts) == 1:
+            return parts[0]
+        return {'$and': parts}
+    result = _validate_fragment(q, type_map)
+    return result if result is not None else {}
+
+
 def validate_query_fields(query):
     """Validate field names in a MongoDB-style query against known output types.
 
@@ -372,80 +453,8 @@ def validate_query_fields(query):
     if not query or not isinstance(query, dict):
         return query
 
-    from secator.output_types import OUTPUT_TYPES
-
-    type_map = {
-        cls.__dataclass_fields__['_type'].default: cls
-        for cls in OUTPUT_TYPES
-        if '_type' in getattr(cls, '__dataclass_fields__', {})
-        and isinstance(cls.__dataclass_fields__['_type'].default, str)
-    }
-
-    def _validate_fragment(fragment):
-        """Validate a single query fragment against known output type fields.
-
-        Returns None if all user-specified fields were invalid (signals the caller
-        to drop this fragment from an $or/$and list rather than keeping a bare
-        {'_type': 'x'} that would match every object of that type).
-        """
-        if not isinstance(fragment, dict):
-            return fragment
-        _type = fragment.get('_type')
-        if not _type:
-            return fragment
-        cls = type_map.get(_type)
-        if cls is None:
-            return fragment
-        valid_fields = set(cls.fields())
-        result = {}
-        user_fields_seen = 0
-        user_fields_kept = 0
-        for k, v in fragment.items():
-            if k.startswith('$') or k.startswith('_'):
-                result[k] = v
-                continue
-            user_fields_seen += 1
-            top_field = k.split('.')[0]
-            if top_field in valid_fields:
-                result[k] = v
-                user_fields_kept += 1
-            else:
-                from secator.output_types import Warning
-                public_fields = sorted(f for f in valid_fields if not f.startswith('_'))
-                console.print(Warning(
-                    message=f"Field '{k}' does not exist on type '{_type}'. "
-                            f"Available fields: {', '.join(public_fields)}"
-                ))
-        # If the fragment had user-specified fields but ALL were invalid, signal
-        # the caller to drop this fragment entirely rather than keeping a bare
-        # {'_type': _type} that would match every object of that type.
-        if user_fields_seen > 0 and user_fields_kept == 0:
-            return None
-        return result
-
-    def _validate(q):
-        if not isinstance(q, dict):
-            return q
-        if '$or' in q:
-            parts = [_validate(sub) for sub in q['$or']]
-            parts = [p for p in parts if p]
-            if len(parts) == 0:
-                return {}
-            if len(parts) == 1:
-                return parts[0]
-            return {'$or': parts}
-        if '$and' in q:
-            parts = [_validate(sub) for sub in q['$and']]
-            parts = [p for p in parts if p]
-            if len(parts) == 0:
-                return {}
-            if len(parts) == 1:
-                return parts[0]
-            return {'$and': parts}
-        result = _validate_fragment(q)
-        return result if result is not None else {}
-
-    return _validate(query)
+    type_map = _build_type_map()
+    return _validate_query(query, type_map)
 
 
 def query_has_type_constraint(query):

--- a/secator/query/utils.py
+++ b/secator/query/utils.py
@@ -3,7 +3,7 @@
 import json
 import re
 
-from secator.output_types import Error, Warning
+from secator.output_types import Error
 from secator.rich import console
 from secator.utils import debug
 
@@ -410,6 +410,7 @@ def validate_query_fields(query):
                 result[k] = v
                 user_fields_kept += 1
             else:
+                from secator.output_types import Warning
                 public_fields = sorted(f for f in valid_fields if not f.startswith('_'))
                 console.print(Warning(
                     message=f"Field '{k}' does not exist on type '{_type}'. "
@@ -427,7 +428,7 @@ def validate_query_fields(query):
             return q
         if '$or' in q:
             parts = [_validate(sub) for sub in q['$or']]
-            parts = [p for p in parts if p is not None]
+            parts = [p for p in parts if p]
             if len(parts) == 0:
                 return {}
             if len(parts) == 1:
@@ -435,7 +436,7 @@ def validate_query_fields(query):
             return {'$or': parts}
         if '$and' in q:
             parts = [_validate(sub) for sub in q['$and']]
-            parts = [p for p in parts if p is not None]
+            parts = [p for p in parts if p]
             if len(parts) == 0:
                 return {}
             if len(parts) == 1:

--- a/tests/unit/test_query_utils.py
+++ b/tests/unit/test_query_utils.py
@@ -5,6 +5,7 @@ from secator.query.utils import (
     python_expr_to_mongo,
     validate_query_fields,
     query_has_type_constraint,
+    _warn_unknown_field,
 )
 
 
@@ -302,14 +303,14 @@ class TestValidateQueryFields:
         # When ALL user-specified fields are invalid, the fragment is dropped entirely
         # (returning {} rather than {'_type': 'technology'} which would match everything).
         q = {'_type': 'technology', 'name': {'$regex': '(HSTS|php)'}}
-        with patch('secator.query.utils.console') as mock_console:
+        with patch('secator.query.utils._warn_unknown_field') as mock_warn:
             result = validate_query_fields(q)
         assert result == {}
-        mock_console.print.assert_called_once()
-        warning_obj = mock_console.print.call_args[0][0]
-        assert 'name' in warning_obj.message
-        assert 'technology' in warning_obj.message
-        assert 'product' in warning_obj.message  # one of the available fields
+        mock_warn.assert_called_once()
+        field_name, type_name, valid_fields = mock_warn.call_args[0]
+        assert field_name == 'name'
+        assert type_name == 'technology'
+        assert 'product' in valid_fields  # one of the available fields
 
     def test_or_validates_each_fragment(self):
         # The technology fragment has no valid user fields, so it is dropped from $or.
@@ -320,10 +321,10 @@ class TestValidateQueryFields:
                 {'_type': 'technology', 'name': {'$regex': '(HSTS|php)'}},
             ]
         }
-        with patch('secator.query.utils.console') as mock_console:
+        with patch('secator.query.utils._warn_unknown_field') as mock_warn:
             result = validate_query_fields(q)
         assert result == {'_type': 'url', 'status_code': {'$ne': 200}}
-        mock_console.print.assert_called_once()
+        mock_warn.assert_called_once()
 
     def test_or_with_partial_invalid_fields(self):
         # When one fragment has some valid and some invalid fields, only the invalid
@@ -334,7 +335,7 @@ class TestValidateQueryFields:
                 {'_type': 'technology', 'product': 'nginx', 'name': 'bogus'},
             ]
         }
-        with patch('secator.query.utils.console') as mock_console:
+        with patch('secator.query.utils._warn_unknown_field') as mock_warn:
             result = validate_query_fields(q)
         assert result == {
             '$or': [
@@ -342,7 +343,7 @@ class TestValidateQueryFields:
                 {'_type': 'technology', 'product': 'nginx'},
             ]
         }
-        mock_console.print.assert_called_once()
+        mock_warn.assert_called_once()
 
     def test_and_validates_each_fragment(self):
         # The technology fragment (all user fields invalid) is dropped from $and.
@@ -353,10 +354,10 @@ class TestValidateQueryFields:
                 {'_type': 'technology', 'name': {'$regex': 'php'}},
             ]
         }
-        with patch('secator.query.utils.console') as mock_console:
+        with patch('secator.query.utils._warn_unknown_field') as mock_warn:
             result = validate_query_fields(q)
         assert result == {'_context.scan_id': '5'}
-        mock_console.print.assert_called_once()
+        mock_warn.assert_called_once()
 
     def test_unknown_type_passes_through(self):
         q = {'_type': 'unknown_type', 'foo': 'bar'}
@@ -377,10 +378,10 @@ class TestValidateQueryFields:
     def test_multiple_invalid_fields_all_warned(self):
         # All user fields invalid → fragment dropped entirely, returning {}
         q = {'_type': 'technology', 'name': 'php', 'bogus': 'x'}
-        with patch('secator.query.utils.console') as mock_console:
+        with patch('secator.query.utils._warn_unknown_field') as mock_warn:
             result = validate_query_fields(q)
         assert result == {}
-        assert mock_console.print.call_count == 2
+        assert mock_warn.call_count == 2
 
     def test_valid_url_status_code(self):
         q = {'_type': 'url', 'status_code': {'$ne': 200}}

--- a/tests/unit/test_query_utils.py
+++ b/tests/unit/test_query_utils.py
@@ -1,11 +1,9 @@
-from unittest.mock import patch
 from secator.query.utils import (
     expand_runner_paths,
     parse_report_paths,
     python_expr_to_mongo,
     validate_query_fields,
     query_has_type_constraint,
-    _warn_unknown_field,
 )
 
 
@@ -302,62 +300,67 @@ class TestValidateQueryFields:
     def test_invalid_field_removed_with_warning(self):
         # When ALL user-specified fields are invalid, the fragment is dropped entirely
         # (returning {} rather than {'_type': 'technology'} which would match everything).
+        from secator.rich import console
+        console.export_text(clear=True)
         q = {'_type': 'technology', 'name': {'$regex': '(HSTS|php)'}}
-        with patch('secator.query.utils._warn_unknown_field') as mock_warn:
-            result = validate_query_fields(q)
+        result = validate_query_fields(q)
         assert result == {}
-        mock_warn.assert_called_once()
-        field_name, type_name, valid_fields = mock_warn.call_args[0]
-        assert field_name == 'name'
-        assert type_name == 'technology'
-        assert 'product' in valid_fields  # one of the available fields
+        recorded = console.export_text()
+        assert "Field 'name' does not exist on type 'technology'" in recorded
+        assert 'product' in recorded  # one of the available fields
 
     def test_or_validates_each_fragment(self):
         # The technology fragment has no valid user fields, so it is dropped from $or.
         # The $or collapses to a single item, which is unwrapped.
+        from secator.rich import console
+        console.export_text(clear=True)
         q = {
             '$or': [
                 {'_type': 'url', 'status_code': {'$ne': 200}},
                 {'_type': 'technology', 'name': {'$regex': '(HSTS|php)'}},
             ]
         }
-        with patch('secator.query.utils._warn_unknown_field') as mock_warn:
-            result = validate_query_fields(q)
+        result = validate_query_fields(q)
         assert result == {'_type': 'url', 'status_code': {'$ne': 200}}
-        mock_warn.assert_called_once()
+        recorded = console.export_text()
+        assert "Field 'name' does not exist on type 'technology'" in recorded
 
     def test_or_with_partial_invalid_fields(self):
         # When one fragment has some valid and some invalid fields, only the invalid
         # field is removed; the fragment itself is kept.
+        from secator.rich import console
+        console.export_text(clear=True)
         q = {
             '$or': [
                 {'_type': 'url', 'status_code': {'$ne': 200}},
                 {'_type': 'technology', 'product': 'nginx', 'name': 'bogus'},
             ]
         }
-        with patch('secator.query.utils._warn_unknown_field') as mock_warn:
-            result = validate_query_fields(q)
+        result = validate_query_fields(q)
         assert result == {
             '$or': [
                 {'_type': 'url', 'status_code': {'$ne': 200}},
                 {'_type': 'technology', 'product': 'nginx'},
             ]
         }
-        mock_warn.assert_called_once()
+        recorded = console.export_text()
+        assert "Field 'name' does not exist on type 'technology'" in recorded
 
     def test_and_validates_each_fragment(self):
         # The technology fragment (all user fields invalid) is dropped from $and.
         # The $and collapses to a single item, which is unwrapped.
+        from secator.rich import console
+        console.export_text(clear=True)
         q = {
             '$and': [
                 {'_context.scan_id': '5'},
                 {'_type': 'technology', 'name': {'$regex': 'php'}},
             ]
         }
-        with patch('secator.query.utils._warn_unknown_field') as mock_warn:
-            result = validate_query_fields(q)
+        result = validate_query_fields(q)
         assert result == {'_context.scan_id': '5'}
-        mock_warn.assert_called_once()
+        recorded = console.export_text()
+        assert "Field 'name' does not exist on type 'technology'" in recorded
 
     def test_unknown_type_passes_through(self):
         q = {'_type': 'unknown_type', 'foo': 'bar'}
@@ -377,11 +380,14 @@ class TestValidateQueryFields:
 
     def test_multiple_invalid_fields_all_warned(self):
         # All user fields invalid → fragment dropped entirely, returning {}
+        from secator.rich import console
+        console.export_text(clear=True)
         q = {'_type': 'technology', 'name': 'php', 'bogus': 'x'}
-        with patch('secator.query.utils._warn_unknown_field') as mock_warn:
-            result = validate_query_fields(q)
+        result = validate_query_fields(q)
         assert result == {}
-        assert mock_warn.call_count == 2
+        recorded = console.export_text()
+        assert "Field 'name' does not exist on type 'technology'" in recorded
+        assert "Field 'bogus' does not exist on type 'technology'" in recorded
 
     def test_valid_url_status_code(self):
         q = {'_type': 'url', 'status_code': {'$ne': 200}}

